pkg/store/postings_codec.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"bytes"
+
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/encoding"
+	"github.com/prometheus/prometheus/tsdb/index"
+)
+
+const (
+	// These headers should not be a prefix of each other.
+	codecHeaderRaw    = "diff+varint+raw"
+	codecHeaderSnappy = "diff+varint+snappy"
+)
+
+// isDiffVarintEncodedPostings returns true, if input looks like it has been encoded by diff+varint(+snappy) codec.
+func isDiffVarintEncodedPostings(input []byte) bool {
+	return bytes.HasPrefix(input, []byte(codecHeaderRaw)) || bytes.HasPrefix(input, []byte(codecHeaderSnappy))
+}
+
+// diffVarintEncode encodes postings into diff+varint representation and optional snappy compression.
+func diffVarintEncode(p index.Postings, useSnappy bool) ([]byte, error) {
+	buf := encoding.Encbuf{}
+
+	// If we're returning raw data, write the header to the buffer, and then return buffer directly.
+	if !useSnappy {
+		buf.PutString(codecHeaderRaw)
+	}
+
+	prev := uint64(0)
+	for p.Next() {
+		v := p.At()
+		if v < prev {
```

Comment:
What if v i `0`?

Reply:
I *believe* that contract of `index.Postings` is that values are increasing [it is not explicit about it, although the file format is and much of the code using postings requires this property]. That implies that `p.At()` can only be 0 as the very first value, which is ok (`prev` is 0 as well at that point, and `v < prev` is not true).

Reply:
fair point. Thanks